src/mainloop.cc

```diff
@@ -0,0 +1,64 @@
+#include "mainloop.h"
+
+#include <uv.h>
+#include <glib.h>
+
+using namespace v8;
+
+struct ThreadData {
+  uv_thread_t thread_handle;
+  uv_mutex_t mutex_handle;
+  uv_async_t async_handle;
+  GPollFD *fds;
+  gint nfds;
+};
+
+static void DispatchGLibMainloop (uv_async_t *async_handle) {
```

Comment:
Sure. Need to update the README too with this from end user perspective.

So this is just for using purely from nodejs (not electron, where chromium uses glib's mainloop to power its event loop called MessageLoop).

The idea is to drive the usual `while (running) { poll(fds, timeout); }` glib mainloop on a separate thread from libuvs mainloop, because glibs mainloop is not easily embedable. But we still want to be able to have our glib async functions wake up on the main thread so they can calll back into v8s javascript context. So we trigger the specific part of the mainloop that will actually distpach events (`g_main_context_dispatch`) back on libuv's main thread.

Something like
```
glib thread: poll for events
glib thread: got events. wakeup libuv thread and lock
libuv thread: dispatch glib events
libuv thread: unlock glib thread
glib thread: poll for events
...
```

Credit to Jasper for the approach.

Long term, if glib moves to epoll (which newer event loops use on linux), we could just poll on a single fd from libuv and drive the glib mainloop in the same thread. Possible with the current state of things, but it would have worked out to a lot more code.

I'll put an abbreviated version of this in the comments.